### PR TITLE
New article for Blazor CSS isolation

### LINKED ISSUE #19360: [RC1] Blazor CSS isolation
* Add *.razor.css files side by side to include scoped css styles.
* Use the `::deep` token in your razor.css file to limit the applicability of the scope isolation so you can target markup defined by other components within your scope.
* Scoped styles are available under `_framework/scoped.styles.css`
* Document how automatic bundling can be disabled to use a separate third-party tool for post-processing the scoped css.
  * This area might receive small tweaks, so hold off on this.
* Document how to pre-process the css files if you are using another pre-processor like LESS or SASS.
  * This might undergo some tweaks too.

We'll be providing more content here before this is ready.

/cc: @SteveSandersonMS

### PR DESCRIPTION
**Pull request by daveabrock**
Here's the new article on Blazor CSS isolation. I'm looking forward to your feedback. 

I struggled a little with the CSS preprocessor section and how much in detail we should go into it. After all, CSS isolation works out of the box well, so don't want to overwhelm, but there's also a lot of folks that are interested in this. 

cc @guardrex @javiercn 

Fixes #19360

### CODE REVIEW

**Comment by guardrex** [comment]
We haven't forgotten about this @daveabrock ... we should be getting to it *very soon*.

File: aspnetcore/blazor/components/css-isolation.md
Comment (javiercn) [inline aspnetcore/blazor/components/css-isolation.md:125]: `CopyToOutputDirectory` this seems wrong.

File: aspnetcore/blazor/components/css-isolation.md
Comment (meziantou) [inline aspnetcore/blazor/components/css-isolation.md:75]: In this case the selector applies to the Counter and CounterChild components because the `h1`s are not at the top level, but are wrapped into a `div`. `::deep` only selects elements that are descendants of an element with the generated attribute (`[b-...]`).

````razor
@page "/counter"
<h1>Not styled by scoped CSS with ::deep</h1>
<div>
    <h1>Styled by scoped CSS</h1>
</div>
````

````html
<!-- [b-a92ooaz332] h1 doesn't match this element -->
<h1 b-a92ooaz332>Not styled by scoped CSS with ::deep</h1> 
<div b-a92ooaz332>
    <h1 b-a92ooaz332>Styled by scoped CSS</h1>
</div>
````

`h1, ::deep h1` would select any `h1` in both components. I think this behavior should be documented somewhere.

Comment (daveabrock) [reply aspnetcore/blazor/components/css-isolation.md:75]: I actually found that if you do this, the parent component gets styled `red` and the children `brown`...

```css
::deep h1 {
    color: red;
}

h1 {
    color: brown;
}
```

...and precedence matters! That is, if you do `h1` first and `::deep h1` second, the parent and child selector is both red.

Then to make them all red, `::deep h1` works without the need to do `h1, ::deep h1`. I would appreciate if you gave this a shot to prove I'm not (too) crazy. 😀

Comment (meziantou) [reply aspnetcore/blazor/components/css-isolation.md:75]: The declaration order should not matter here. `::deep h1` is translated to `[b-xxx] h1`. This selector is more specific ([specificity](https://developer.mozilla.org/en-US/docs/Web/CSS/Specificity)) than `h1`, so `::deep h1` overrides `h1` when both selectors apply to an element.

Using the Counter and CounterChild components, the h1 is brown in the parent component, and red in the child component. It's the same if you change the declaration order.

Let's have fun with the `::deep` pseudo-selector 😊

For all examples 
- the style is `::deep h1 { color: red; }`
- the child component is `<h1>Child component</h1>`

### Example 1

````
<h1>Counter</h1>
<div>
    <CounterChild />
</div>
````

Parent h1: unstyled. Child h1: red.

### Example 2

````
<h1>Counter</h1>
<CounterChild />
````

Parent h1: unstyled. Child h1: unstyled.

### Example 3

````
<div>
    <h1>Counter</h1>
    <CounterChild />
</div>
````

Parent h1: red. Child h1: red.


### Example 4

````
<div>
    <h1>Counter</h1>
</div>
<CounterChild />
````

Parent h1: red. Child h1: unstyled.

> "To apply changes to a child component, use the `::deep` combinator in the parent component's `razor.css` file."

This doesn't seem to be true for example 2 or 4. If you can validate my examples, I think it would make sense to add a note that a `div` is required for the `::deep` selector to work as expected.

Comment (meziantou) [reply aspnetcore/blazor/components/css-isolation.md:75]: I've edited the first part of my previous comment because I misread your comment :(

Comment (daveabrock) [reply aspnetcore/blazor/components/css-isolation.md:75]: Thanks for the great clarificationsI I'll make my updates tonight after the day job and push to the PR.

File: aspnetcore/blazor/components/css-isolation.md
Comment (daveabrock) [reply aspnetcore/blazor/components/css-isolation.md]: @javiercn - definitely is. I replaced `CopyToOutputDirectory` to `CssScoped` in my latest push. Apologies for the brain fart (probably muscle memory from years of `CopyToOutputDirectory`).

File: aspnetcore/blazor/components/css-isolation.md
Comment (daveabrock) [reply aspnetcore/blazor/components/css-isolation.md]: Updated the docs with my latest push. Thanks again for the great feedback!

**Comment by daveabrock** [comment]
@javiercn and @guardrex - updated the docs based on @javiercn and @meziantou feedback. If that is all good, @guardrex might be able to take a look now...

**Comment by guardrex** [comment]
I'll wait for @javiercn to sign off first ... then make a pass and ping u for a final look.

File: aspnetcore/blazor/components/css-isolation.md
Comment (meziantou) [reply aspnetcore/blazor/components/css-isolation.md]: The latest version is very clear! Good job :)

**Comment by guardrex** [comment]
@captainsafia ... Can you perform the technical review and then I'll round up the UE updates?

File: aspnetcore/blazor/components/css-isolation.md
Comment (captainsafia) [inline aspnetcore/blazor/components/css-isolation.md:99]:
```suggestion
> The `::deep` combinator works only with descendant elements. The following HTML structure applies the `h1` styles to components as expected:
```

And maybe include the CSS associated with this explanation?

File: aspnetcore/blazor/components/css-isolation.md
Comment (captainsafia) [inline aspnetcore/blazor/components/css-isolation.md:128]: This first sentence feels a little weird in conjunction with the next. Maybe say something like "CSS isolation is designed to work out-of-the-box, but provides configuration for some advanced scenarios..."

File: aspnetcore/blazor/components/css-isolation.md
Comment (captainsafia) [inline aspnetcore/blazor/components/css-isolation.md:140]: I would clarify that this is specific to CSS generated from `MyComponent.razor.css`.

Comment (daveabrock) [reply aspnetcore/blazor/components/css-isolation.md:140]: Changed to:

In this example, the CSS generated from `MyComponent.Razor.css` changes its scope identifier from `b-<10-character-string>` to `my-custom-scope-identifier`.

File: aspnetcore/blazor/components/css-isolation.md
Comment (daveabrock) [reply aspnetcore/blazor/components/css-isolation.md:128]: Good point. Changed to:

CSS isolation is designed to work out-of-the-box, but provides configuration for some advanced scenarios—such as when there are dependencies on existing tools or workflows.

File: aspnetcore/blazor/components/css-isolation.md
Comment (daveabrock) [reply aspnetcore/blazor/components/css-isolation.md:22]: Changed to:

For a `MyComponent` that has a `MyComponent.razor` file, create a file alongside your component called `MyComponent.razor.css`. The `MyComponent` value in the `razor.css` filename is **not** case-sensitive.

File: aspnetcore/blazor/components/css-isolation.md
Comment (daveabrock) [reply aspnetcore/blazor/components/css-isolation.md:16]: Changed to:

CSS isolation simplifies an app's CSS footprint by preventing dependencies on global styles, and also helps to avoid styling conflicts between other components or libraries. Use CSS isolation to only apply styles to the current component.

File: aspnetcore/blazor/components/css-isolation.md
Comment (daveabrock) [reply aspnetcore/blazor/components/css-isolation.md:88]: Updated confusion between the format `MyComponent.razor.css` and the example `ChildExample.razor.css`.

"By default, CSS isolation only applies to the component you associate with the format `MyComponent.razor.css`... The following example shows a parent component, called `ChildExample` in a file called `ChildExample.razor`, with a child component called `MyChild`."

File: aspnetcore/blazor/components/css-isolation.md
Comment (daveabrock) [reply aspnetcore/blazor/components/css-isolation.md:96]: Mostly cleaned up from other comment, but changed to:

"The `h1` style now applies to the parent `ChildExample` and `FirstChild` component without the need to create separate scoped CSS files for the child component."

**Comment by daveabrock** [comment]
Pushed changes based on recent feedback (thanks @captainsafia). cc @guardrex

**Comment by guardrex** [comment]
Cool ... I'll make an edit pass asap, but it might be early next week. I'm 🏃🏃🏃😅 on Blazor security work at the moment.

**Comment by danroth27** [comment]
@daveabrock @guardrex FYI - there are various changes coming to CSS isolation in .NET 5 RC2 that we'll need to react to in this doc: https://github.com/dotnet/aspnetcore/pull/25565

**Comment by guardrex** [comment]
@daveabrock ... I'll hold off on my pass until you get a chance to react; or if you're too busy for the updates, I'll take a shot at getting them into the doc.

**Comment by daveabrock** [comment]
> @daveabrock ... I'll hold off on my pass until you get a chance to react; or if you're too busy for the updates, I'll take a shot at getting them into the doc.

I can take a look at these. I'll do some updates this weekend, then test the updates when RC2 comes out. Which I'm sure is *very soon*. 😁

**Comment by daveabrock** [comment]
With RC2 out now, @guardrex, should have updates in tonight and updates ready for review tomorrow AM.

**Comment by guardrex** [comment]
Sure thing ... no worries ... we're still fine here. We target to make sure that docs are up by release at the latest, then we go for preview when we're able. I certainly wouldn't rush you on updates given that you're putting in your time and effort on this. It's great to have your help. I'm **_still_** getting the last of the .NET Blazor WASM security updates in, so you've taken a real burden off of me so that I can focus on these security updates. If you even need a little more time, don't sweat it ... take the time. We still have a few weeks left to go.

**Comment by daveabrock** [comment]
OK, @guardrex, I pushed updates for RC2. I created a Razor Class Library in the sample app to understand the changes and how projects are referenced—slick!

A subtle change is in the first note, I added the "Razor" word:
"In order to guarantee style isolation when bundling occurs, @import Razor blocks are not supported with scoped CSS files."

I clarified this because the generated bundles use CSS imports but this is still not allowed:

```razor
@css { }
```

Adding "Razor" to the note probably avoids further confusion.

**Comment by guardrex** [comment]
Thanks @daveabrock ... I'll be free I think on ...... errrrrrr 🤔 ..... Monday late morning. My Friday is already full, and I have a priority WASM hosting regression to address Monday morning.

**Comment by daveabrock** [comment]
@guardrex - sounds great! (The review, not the regression. 😉)

**Comment by guardrex** [comment]
@daveabrock ... **_Almost here!_** 🏃😅 

It *only* took **TEN** hours today to get my hosted WASM with AAD groups and roles security code running. **_Yikes!_** :smile: Well .... at least it **_IS_** running! :tada::beers:

This PR is coming up next. I reached a 🧠🔥 *brain fry* 🧠🔥 about an hour ago (and I need to get the Blazor 5.0 migration bits into the *Migration to 5.0* topic tonight), so I won't get to this by EOD. Hopefully, I'll reach this by tomorrow (Tuesday).

**Comment by daveabrock** [comment]
> It _only_ took **TEN** hours today to get my hosted WASM with AAD groups and roles security code running. **_Yikes!_** 😄 Well .... at least it **_IS_** running! 🎉🍻

Thoughts and prayers. I really appreciate you keeping me up to date!

**Comment by guardrex** [comment]
Sure thing. I don't think they'll be much more for you to do. I'll make a direct edit pass that will probably consist of nit updates for our style guidelines. I'll ping you to review the updates, then we'll ask the product unit if it's ok to get this content merged.

**Comment by guardrex** [comment]
Great job ... I ❤️ the doc!

You might (er ... *likely will*) need to edit my changes a bit. I may have 💥:cry: a bit of meaning that you'll need to correct.

Notes on my changes:

* I use placeholders in the format `{PLACEHOLDER}` with text that explains what the placeholder is. This is because localization doesn't happen on code. Therefore, we have a better chance that non-English readers will understand the docs.
* Your Parent-Child example ... I changed it to "Parent" and "Child" because having the word "child" in the name of the parent component didn't seem valuable at all and may lead to confusion. I also took off the `@page` directive from the child as unnecessary. See if that example holds up with my changes. If not, change it, but try not to use "child" in the parent component's name.
* There's a friendly disagreement among doc authors on this repo about prepositional phrases and commas in openers. IMO, a prepositional phrase that modifies an object of another prepositional phrase is almost always **_an essential phrase_**; therefore, I personally almost **never** comma them. Example: "For example to eat the cake, I grabbed a fork." Authors here feel that the second phrase is parenthetical and will set it off with commas ("For example, to eat the cake, I grabbed a fork."). I don't think that's correct. IMO, the 2nd phrase specifies the word "example" exactly and thus is necessary/essential. That's almost always the case with a prepositional phrase that follows another prepositional phrase. Anyway ... to solve it and make everyone happy, I just try to avoid the construction. [btw- This is also true with adverbs and prepositional phrases. Example: "Additionally to eat the cake, I use a fork." (no comma)] You'll see some updates on the commit along these lines. Often, "for example" can just be eliminated as unnecessary. Other times, the sentence can be changed to avoid double-prepositional phrase openers. .......... Ok .... :smile: ... *Professor Luke's English 101 course is now OVER!* 😁 lol

If you find that you're getting too many pings from being in the `author` metadata after this is published, we can assign it to me for the pings. That might save your Inbox a bit if you're getting too many hits from readers. However, it isn't a bad idea to leave yourself there for a while going into early next year so that you can interact immediately with anyone who opens an issue on the topic. I leave this aspect up to you.

After you **_fix_** any of my edits that went sideways, we'll see about wrapping this up and getting it merged.

**Comment by daveabrock** [comment]
Actually, I went though this when I was emailed about your push and I think it looks good, and you haven't lost any meaning with your changes.

With this and my blog post being the top search result for this topic, for better or worse I am the "Blazor CSS isolation guy" so I may as well embrace it! We can leave my name on, and I'll ping you if it gets to be a little too much.

**Comment by guardrex** [comment]
"*I may as well embrace it!*" ..... put that on my tombstone please. ⚰️ lol

If you end up with too many pings and having to deal with too many dev support requests, just send a PR changing the `author` metadata to `guardrex` and leave the author byline alone. You always keep the credit ... I'll take the **_HITS!_** 👦🤛😵 *ouch!*

@captainsafia ... Do you want to take a final look and then I can merge this, or do you want to call for additional product unit review before merging? I suspect that all of the GA coverage will be reviewed in the live preview topics in the days leading up to release or very soon thereafter, so I think we can go ahead and merge this.

File: aspnetcore/blazor/components/css-isolation.md
Comment (captainsafia) [inline aspnetcore/blazor/components/css-isolation.md:24]:
```suggestion
For example to add CSS isolation to the `Counter` component in the default Blazor project template, add a new file named `Counter.razor.css` along side the `Counter.razor` file, then add the following CSS:
```

File: aspnetcore/blazor/components/css-isolation.md
Comment (captainsafia) [inline aspnetcore/blazor/components/css-isolation.md:147]: Is this the right place to include this heading?

Comment (daveabrock) [reply aspnetcore/blazor/components/css-isolation.md:147]: Yeah, there's no natural way unless we include it in the main section where it's discussed. I personally wanted to save the config to the end as it is optional and probably rarely used.

**Comment by guardrex** [comment]
btw @daveabrock ... For that Line 24 change ... "along side" 👉 "alongside" (one word)

**Comment by daveabrock** [comment]
🚢?!?!?

**Comment by daveabrock** [comment]
🚢?!?!?!

**Comment by guardrex** [comment]
> 🚢?!?!?

We'll wait for @captainsafia to give a final *okey-dokey* and say if we should call anyone else in. Personally, I think it's good to go. We can have additional :eyes: anytime up to and just after 5.0 GA.

**Review by captainsafia — APPROVED** [review]

**Comment by javiercn** [comment]
I'll do a pass on this tomorrow

**Review by javiercn — APPROVED** [review]
Looks great!

**Comment by guardrex** [comment]
Thank U so much @daveabrock for taking this on. Your topic contribution is 🥇 **_greatly appreciated!_** 🙇‍♂️.

After merging this PR, I'll do a live docs update, so this should be online shortly in the preview docs.

**Comment by voroninp** [comment]
Still cannot locate sections about minification.

**Pull request merged**
